consensus/src/quorum_store/types.rs

```diff
@@ -72,143 +32,157 @@ impl PersistedValue {
+        self.txns
     }
 
-    pub fn batch_id(&self) -> BatchId {
-        self.batch_id
+    pub fn num_txns(&self) -> usize {
+        self.txns.len()
     }
 
-    pub fn maybe_expiration(&self) -> Option<LogicalTime> {
-        self.maybe_expiration
+    pub fn num_bytes(&self) -> usize {
+        *self
+            .num_bytes
+            .get_or_init(|| bcs::serialized_size(&self).expect("unable to serialize batch payload"))
     }
 }
-
 #[derive(Clone, Debug, Deserialize, Serialize)]
-pub struct Fragment {
-    source: PeerId,
-    fragment_info: FragmentInfo,
+pub struct Batch {
+    batch_info: BatchInfo,
+    payload: BatchPayload,
 }
 
-impl Fragment {
+impl Batch {
     pub fn new(
-        epoch: u64,
         batch_id: BatchId,
-        fragment_id: usize,
-        fragment_payload: Vec<SerializedTransaction>,
-        maybe_expiration: Option<LogicalTime>,
-        peer_id: PeerId,
+        payload: Vec<SignedTransaction>,
+        expiration: LogicalTime,
+        batch_author: PeerId,
     ) -> Self {
-        let fragment_info = FragmentInfo::new(
-            epoch,
+        let payload = BatchPayload::new(payload);
+        let batch_info = BatchInfo::new(
+            batch_author,
             batch_id,
-            fragment_id,
-            fragment_payload,
-            maybe_expiration,
+            expiration,
+            payload.hash(),
+            payload.num_txns() as u64,
+            payload.num_bytes() as u64,
         );
         Self {
-            source: peer_id,
-            fragment_info,
+            batch_info,
+            payload,
         }
     }
 
-    pub fn verify(&self, peer_id: PeerId) -> anyhow::Result<()> {
-        if let Some(expiration) = &self.fragment_info.maybe_expiration() {
-            if expiration.epoch() != self.fragment_info.epoch {
-                return Err(anyhow::anyhow!(
-                    "Epoch mismatch: info: {}, expiration: {}",
-                    expiration.epoch(),
-                    self.fragment_info.epoch
-                ));
-            }
-        }
-        if self.source == peer_id {
-            Ok(())
-        } else {
-            Err(anyhow::anyhow!(
-                "Sender mismatch: peer_id: {}, source: {}",
-                self.source,
-                peer_id
-            ))
-        }
+    pub fn verify(&self) -> anyhow::Result<()> {
+        ensure!(
+            self.payload.hash() == self.batch_info.digest,
+            "Payload hash doesn't match the digest"
+        );
+        ensure!(
+            self.payload.num_txns() as u64 == self.batch_info.num_txns,
+            "Payload num txns doesn't match batch info"
+        );
+        ensure!(
+            self.payload.num_bytes() as u64 == self.batch_info.num_bytes,
+            "Payload num bytes doesn't match batch info"
+        );
+        Ok(())
     }
 
-    pub fn epoch(&self) -> u64 {
-        self.fragment_info.epoch
+    pub fn digest(&self) -> HashValue {
```

Comment:
In fact, this is manually implemented for Batch struct but not for others, this is exactly the interfaces I'd make a derivable trait for all structs that include info: BatchInfo